### ISSUE TITLE
[CI] only build/test images that have been updated by the commit

### ISSUE THREAD

**Issue by vince-fugnitto** [theia-ide/theia-apps#214]
**Description**

Only build and test image(s) which are directly affected by the commit.

**Problem**

For each pull-request `travis` builds and tests every single image in the repository irregardless of if the image has actually been updated which is both **time and resource consuming**.

**Solution**

The proposal is to include [**`conditional builds/stages`**](https://docs.travis-ci.com/user/conditional-builds-stages-jobs/) to determine which image(s) should be built and tested.

The idea is to do so by either of the following [**`methods`**](https://docs.travis-ci.com/user/conditions-v1):
1. add a `tag` which can be set through a post-commit hook
2. make use of the `commit-message` which can include which images are updated and be a standard part of the review process.

**Comment by marcdumais-work** [comment]
(test - please ignore)

@theia-ide/ip-help testing the new team in a mention

**Comment by dwjbosman** [comment]
Wouldn't it be possible to:
1. Use git to determine which files were changed
2. Based on the changed files determine the application folders and theia apps that have changed.
3. Then, based on the changed theia apps, run the build/test scripts.

That way a user wouldn't have to do anything special. 

A bash script should be able to do step 1 and 2. Call the script from Travis.

**Comment by dwjbosman** [comment]
Hi, I tried to address this using the TRAVIS_EVENT_TYPE environment variable. Please see the pull request comments for more info.

**Comment by marcdumais-work** [comment]
Hi @dwjbosman ,

I think it would be simpler for both us and you, if we were to give you write access to this repo here; this way your PRs can originate from the main repo and be covered by the GitHub token, reducing build failures. WDYT?

**Comment by dwjbosman** [comment]
That's fine. Shall I then recreate the pull requests?

**Comment by marcdumais-work** [comment]
@dwjbosman I have added you as a collaborator in this repo - you should receive an email notification with a link to accept

> Shall I then recreate the pull requests?

Yes, I think that's the way to go - thanks.

**Comment by dwjbosman** [comment]
I have not received an activation email (checked the spam etc.). Can you resend it?

**Comment by marcdumais-work** [comment]
> I have not received an activation email (checked the spam etc.). Can you resend it?

The invitation is no longer pending - you are listed as member with write access. Please try pushing your "build_on_changes" branch to the main repo and let me know if it doesn't work. If it does work, everything is good.

**Comment by dwjbosman** [comment]
After merging the build immediately failed :)

There was an error:
ambiguous argument 'HEAD~1...HEAD': unknown revision or path not in the working tree.

Could it be that git clone depth should be 2 in .travis.yml?

1. When committing to pull request a depth of 1 is ok because Travis will do a temporary commit

2. When the build trigger originates from a pull request merge then Travis will not do the temporary commit, and a "depth 1" will result in the problem that HEAD~1 is not available.

**Comment by marcdumais-work** [comment]
@dwjbosman Looks like you're probably right. Can you do a quick PR to try dept of 2?

**Comment by marcdumais-work** [comment]
It looks like the [nightly build](https://travis-ci.org/theia-ide/theia-apps/builds/588823783?utm_source=github_status&utm_medium=notification) has failed as well, but for different reason, in de `deploy` phase: https://travis-ci.org/theia-ide/theia-apps/jobs/588823786#L859-L863

```
Deploying application
invalid reference format
Script failed with status 1
failed to deploy
/home/travis/.travis/functions: line 485:  4194 Terminated              travis_wait 30 sleep 1800  (wd: ~/build/theia-ide/theia-apps)
```

**Comment by marcdumais-work** [comment]
> @dwjbosman Looks like you're probably right. Can you do a quick PR to try dept of 2?

I'll give this a quick try - I think we need to merge the change before we can see if it works anyway, so I'll skip the review and just merge.

**Comment by marcdumais-work** [comment]
Works ok now with `depth=2`: https://travis-ci.org/theia-ide/theia-apps/builds/588906080?utm_source=github_status&utm_medium=notification

**Comment by marcdumais-work** [comment]
> It looks like the [nightly build](https://travis-ci.org/theia-ide/theia-apps/builds/588823783?utm_source=github_status&utm_medium=notification) has failed as well, but for different reason, in the `deploy` phase: https://travis-ci.org/theia-ide/theia-apps/jobs/588823786#L859-L863
> 
> ```
> Deploying application
> invalid reference format
> Script failed with status 1
> failed to deploy
> /home/travis/.travis/functions: line 485:  4194 Terminated              travis_wait 30 sleep 1800  (wd: ~/build/theia-ide/theia-apps)
> ```

Update: the nightly build no longer fails, but it also does not rebuild/test all images, modified in the latest HEAD commit or not, as I think was the intention. See: https://travis-ci.org/theia-ide/theia-apps/builds/590221434?utm_source=github_status&utm_medium=notification

One reason for the nightly build to always redo all images is to detect if there was a breaking change recently committed in the Theia platform or other Theia extensions. I guess it's mostly the `next` images that are useful for that.

**Comment by dwjbosman** [comment]
The check_changes script uses

`if [ "$TRAVIS_EVENT_TYPE" == "api" ]
`
to determine if all should be build. The question is what this value is when cron starts the build....

I just checked the doc. The value is "cron" when executed via cron job thus the script should be changed to

`if [ "$TRAVIS_EVENT_TYPE" != "pull_request" ]`

**Comment by dwjbosman** [comment]
If I see small errors like this. Should I directly change them on master? create a pull request? a branch?

**Comment by marcdumais-work** [comment]
> If I see small errors like this. Should I directly change them on master? create a pull request? a branch?

The repo's config prevents pushing to master without an approved review first, which I think it a good practice. Please create a a PR; it can be reviewed/approved very quickly.

**Comment by dwjbosman** [comment]
See pull request #244

**Comment by vince-fugnitto** [comment]
Fixed by #244 #236
